spanner/client_test.go

```diff
@@ -1435,6 +1421,225 @@ func TestClient_ReadWriteTransactionCommitAlreadyExists(t *testing.T) {
+				}
+				if err != nil {
+					return
+				}
+				var singerID, albumID int64
+				var albumTitle string
+				if err := row.Columns(&singerID, &albumID, &albumTitle); err != nil {
+					return
+				}
+				rowCount++
+			}
+			return
+		}
+		wg.Add(2)
+		go query(&firstTransactionID)
+		go query(&secondTransactionID)
+		wg.Wait()
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if firstTransactionID == nil || secondTransactionID == nil || string(firstTransactionID) != string(secondTransactionID) {
+		t.Fatalf("transactionID mismatch:\nfirst: %v\nsecong: %v", firstTransactionID, secondTransactionID)
+	}
+	requests := drainRequestsFromServer(server.TestSpanner)
+	if err := compareRequests([]interface{}{
+		&sppb.BatchCreateSessionsRequest{},
+		&sppb.ExecuteSqlRequest{},
+		&sppb.ExecuteSqlRequest{},
+		&sppb.CommitRequest{}}, requests); err != nil {
+		t.Fatal(err)
+	}
+	var (
+		callsWithBeginSelector int32
+		callsWithTransactionID int32
+	)
+	for _, req := range requests {
+		if sql, ok := req.(*sppb.ExecuteSqlRequest); ok {
+			if _, ok := sql.Transaction.GetSelector().(*sppb.TransactionSelector_Begin); ok {
+				callsWithBeginSelector++
+			}
+			if _, ok := sql.Transaction.GetSelector().(*sppb.TransactionSelector_Id); ok {
+				callsWithTransactionID++
+			}
+		}
+	}
+	if callsWithBeginSelector != 1 || callsWithTransactionID != 1 {
+		t.Fatal("first statement in concurrent read/write transaction should use TransactionSelector::Begin "+
+			"and others should use transactionID returned from first statement", firstTransactionID, secondTransactionID)
+	}
+}
+
+// Given a transaction, When the first call to ExecuteStreamingSql/StreamingRead returns an UNAVAILABLE error
+// and retry returns Aborted, then the transaction should be retried with an explicit BeginTransaction rpc.
+func TestClient_ReadWriteTransaction_FirstStatementAsQueryReturnsUnavailableRetryReturnsAborted(t *testing.T) {
+	t.Parallel()
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	server.TestSpanner.PutExecutionTime(MethodExecuteStreamingSql,
+		SimulatedExecutionTime{
+			Errors: []error{status.Error(codes.Unavailable, "Temporary unavailable")},
+		})
+	server.TestSpanner.PutExecutionTime(MethodExecuteStreamingSql,
+		SimulatedExecutionTime{
+			Errors: []error{status.Error(codes.Aborted, "Transaction aborted")},
+		})
+	ctx := context.Background()
+	_, err := client.ReadWriteTransaction(ctx, func(ctx context.Context, tx *ReadWriteTransaction) error {
+		iter := tx.Query(ctx, NewStatement(SelectSingerIDAlbumIDAlbumTitleFromAlbums))
+		defer iter.Stop()
+		for {
+			_, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	requests := drainRequestsFromServer(server.TestSpanner)
+	if err := compareRequests([]interface{}{
+		&sppb.BatchCreateSessionsRequest{},
+		&sppb.ExecuteSqlRequest{},
+		&sppb.BeginTransactionRequest{},
```

Comment:
>What happens if the first statement in a transaction returns an error?

@olavloite so is error from first statement is retryable(Aborted/Session not found) we will do explicit begin transaction and retry.